### PR TITLE
Upgrade to relay v7

### PR DESCRIPTION
**Pull request by sorenhoyer**
Adds support for relay 7

Accidentally deleteted the old fork (should have been the DefinitelyTyped one instead!!!) that was linked to the original PR  :clap:, so had to abandon that one, make a new fork and this new PR. For reference, here's the original PR https://github.com/relay-tools/relay-compiler-language-typescript/pull/154, which this one replaces.

Thanks to @renanmav for contributing with a PR and feedback and @alloy @smmoosavi for some great feedback as well.

Related PR's:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39995
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40011
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39665
https://github.com/relay-tools/relay-compiler-language-typescript/pull/154

### CODE REVIEW

**Comment by sorenhoyer** [comment]
Made a PR here for the missing types https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39995

**Comment by sorenhoyer** [comment]
1 more PR for @types/relay-runtime so the following import will work:

`import { ConnectionInterface } from "relay-runtime";`

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40011

File: test/fixtures/type-generator/query-with-defer.graphql [hunk 1]
```diff
@@ -2,16 +2,16 @@ query TestDefer @raw_response_type {
...
+            actor {
+              name
             }
           }
         }
       }
     }
   }
-}
+}
```
Comment (alloy) [inline test/fixtures/type-generator/query-with-defer.graphql]: This fixture currently doesn’t look like the upstream one, what gives? https://github.com/facebook/relay/blob/v7.0.0/packages/relay-compiler/transforms/__tests__/fixtures/relay-defer-stream-transform/query-with-defer.graphql

**Comment by ds300** [comment]
Can confirm that our app is working fine with these changes :+1:

**Comment by alloy** [comment]
@ds300 Great, thanks 👍 

@sorenhoyer Can you update the status of the PR from your perspective?

**Comment by renanmav** [comment]
I'm getting this errors when run the example
```bash
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/ts/app.tsx(59,25):
TS2322: Type '{ readonly " $fragmentRefs": FragmentRefs<"TodoApp_viewer">; }' is not assignable to type '_FragmentRefs<"TodoApp_viewer">'.
  Types of property '' $fragmentRefs'' are incompatible.
    Type 'FragmentRefs<"TodoApp_viewer">' is not assignable to type '"TodoApp_viewer"'.
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/ts/components/TodoApp.tsx(56,21):
TS2322: Type 'TodoApp_viewer' is not assignable to type '_FragmentRefs<"TodoList_viewer">'.
  Types of property '' $fragmentRefs'' are incompatible.
    Type 'FragmentRefs<"TodoListFooter_viewer" | "TodoList_viewer">' is not assignable to type '"TodoList_viewer"'.
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/ts/components/TodoApp.tsx(57,40):
TS2322: Type 'TodoApp_viewer' is not assignable to type '_FragmentRefs<"TodoListFooter_viewer">'.
  Types of property '' $fragmentRefs'' are incompatible.
    Type 'FragmentRefs<"TodoListFooter_viewer" | "TodoList_viewer">' is not assignable to type '"TodoListFooter_viewer"'.
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/ts/components/TodoList.tsx(50,9):
TS2322: Type '{ readonly id: string; readonly complete: boolean | null; readonly " $fragmentRefs": FragmentRefs<"Todo_todo">; }' is not assignable to type '_FragmentRefs<"Todo_todo">'.
  Types of property '' $fragmentRefs'' are incompatible.
    Type 'FragmentRefs<"Todo_todo">' is not assignable to type '"Todo_todo"'.
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/ts/components/TodoList.tsx(51,9):
TS2322: Type 'TodoList_viewer' is not assignable to type '_FragmentRefs<"Todo_viewer">'.
  Types of property '' $fragmentRefs'' are incompatible.
    Type 'FragmentRefs<"Todo_viewer">' is not assignable to type '"Todo_viewer"'.
Version: typescript 3.6.3
Time: 7851ms
```

**Comment by alloy** [comment]
@renanmav That probably means you have multiple copies of the relay-runtime types package.

**Comment by renanmav** [comment]
In fact I had 😆

LGTM

**Comment by sorenhoyer** [comment]
@alloy The status of this PR from my point of view is that I consider this PR ready to merge whenever you do (do you bump the version when you merge or should I make a yarn version bump in a commit?). We're already using an npm pack version of this fork right now at my company, and I also tried setting up a [small app](https://github.com/sorenhoyer/react-relay-boilerplate) to further verify if it seems to be working. So far I've seen no issues.

We're using the new hooks that can be imported from "react-relay/hooks" with relay 7, so I have not tested the old API using the HOC's. Also our app is not very complicated yet, so it would be great if more people can help test in their applications, so we get better coverage.

**Comment by sibelius** [comment]
I think we can ship a new release

and then start working on https://github.com/relay-tools/relay-compiler-language-typescript/pull/149

**Comment by alloy** [comment]
Our Emission app has no hooks and good coverage, so I think we’re good.

So regarding #149, should we merge this PR but not release yet until #149 is merged?

**Comment by sibelius** [comment]
we can release before 149, 149 is not a breaking change

**Comment by alloy** [comment]
Sure, I guess I meant more if it was conceptually part of this overall work, but I’m fine adding that as a minor version 👍

**Comment by alloy** [comment]
Great job on this @sorenhoyer! 🙌

**Comment by alloy** [comment]
:rocket: PR was released in v10.0.0 :rocket:

**Comment by sorenhoyer** [comment]
Should I go ahead and bump the version?

**Comment by dotbear** [comment]
Thanks for your work @sorenhoyer !

**Pull request merged**